src/prefect/engine/cloud/flow_runner.py

```diff
@@ -59,35 +66,136 @@ def say_hello():
+        except Exception:
+            self.logger.exception("Error occured on run")
+
+        self.logger.debug("Flowrunner completed")
+        self._request_exit()
+
+    def run(self, **kwargs: Any) -> State:
+        flow_run_id = prefect.context.get("flow_run_id", "")  # type: str
+
+        if not self.check_valid_initial_state(flow_run_id):
+            raise RuntimeError("Flow run initial state is invalid. It will not be run!")
+
+        # start a state listener thread, pulling states for this flow run id from cloud.
+        # Events are reported back to the main thread (here). Why a separate thread?
+        # Among other reasons, when we start doing subscriptions later, it will continue
+        # to work with little modification (replacing the periodic caller with a thread)
+        self.state_thread = PeriodicMonitoredCall(
+            interval=10,
+            function=self.stream_flow_run_state_events,
+            logger=self.logger,
+            flow_run_id=flow_run_id,
+        )
+        self.state_thread.start(
+            name_prefix="PrefectFlowRunState-{}".format(flow_run_id)
+        )
+
+        # note: this creates a cloud flow runner which has a heartbeat
+        self.worker_thread = threading.Thread(target=self._run, kwargs=kwargs)
+        self.worker_thread.start()
+
+        # handle all flow state events of interest as well as exit requests
         try:
-            # use empty string for testing purposes
-            flow_run_id = prefect.context.get("flow_run_id", "")  # type: str
-            self.client.update_flow_run_heartbeat(flow_run_id)
-            query = {
-                "query": {
-                    with_args("flow_run_by_pk", {"id": flow_run_id}): {
-                        "state": True,
-                        "flow": {"settings": True},
-                    }
+            while True:
+                item = self.queue.get()
+                if item is None:
+                    break
+                elif not isinstance(item, QueueItem):
+                    self.logger.warning("Bad event: {}".format(repr(item)))
+                    continue
+
+                if item.event == "state" and item.payload == Cancelled:
+                    self.cancel()
+                    self._request_exit()
+                elif item.event == "exit":
+                    break
+                else:
+                    self.logger.warning("Unknown event: {}".format(item))
+                    continue
+        except Exception:
+            self.logger.exception("Unhandled exception in the event loop")
+
+        self._on_exit()
+        return self.state
+
+    def fetch_current_flow_run_state(self, flow_run_id: str) -> Type[State]:
+        query = {
+            "query": {
+                with_args("flow_run_by_pk", {"id": flow_run_id}): {
+                    "state": True,
+                    "flow": {"settings": True},
```

Comment:
I don't think this `flow: settings` payload is used anywhere, and especially since this query is periodically issued we could safely remove it to avoid the join. 

```suggestion

```

Reply:
It should be used to determine whether heartbeats should fire: https://github.com/PrefectHQ/prefect/blob/d7b122b2afb7f13265f8d7f0a6ae6d6afd26b634/src/prefect/engine/cloud/flow_runner.py#L83

Reply:
Ah, I missed it. It might be worthwhile to split it into two calls at this time - one for the flow info and one for the flow run info. The join to the flow table is small but unecessary to make on every single poll.

Reply:
Looks like I've removed the ability to disable heartbeats, I'll add that back in. There shouldn't be a reason why we need to get the settings info periodically? That is, we really only need it when we start the flow... the fewer calls the better.